Sources/Networking/Documentation.docc/Documentation.md

```diff
@@ -0,0 +1,297 @@
+    requestAdapters: [RequestAdapting] = [],
+    responseProcessors: [ResponseProcessing] = [StatusCodeProcessor.shared],
+    errorProcessors: [ErrorProcessing] = []
+)
+```
+
+2. Using custom response provider by conforming to ``ResponseProviding``.
+```swift
+init(
+    responseProvider: ResponseProviding,
+    requestAdapters: [RequestAdapting] = [],
+    responseProcessors: [ResponseProcessing] = [StatusCodeProcessor.shared],
+    errorProcessors: [ErrorProcessing] = []
+)
+```
+Adapters and processors are passed during initialisation and cannot be changed afterwards.
+
+There are two methods provided by the ``APIManaging`` protocol:
+
+1. Result is URLSession's default (data, response) tuple.
+```swift
+func request(
+  _ endpoint: Requestable,
+  retryConfiguration: RetryConfiguration?
+) async throws -> Response
+```
+2. Result is custom decodable object.
+```swift
+func request<DecodableResponse: Decodable>(
+  _ endpoint: Requestable,
+  decoder: JSONDecoder,
+  retryConfiguration: RetryConfiguration?
+) async throws -> DecodableResponse
+```
+
+### Example
+In the most simple form, the network request looks like this:
+
+```swift
+try await apiManager.request(UserRouter.getUser)
+```
+
+If you specify object type, the APIManager will automatically perform the decoding (given the received JSON correctly maps to the decodable). You can also specify a custom json decoder.
+
+```swift
+let userResponse: UserResponse = try await apiManager.request(UserRouter.getUser)
+```
+
+Provide a custom after failure ``RetryConfiguration``, specifying the count of retries, delay and a handler that determines whether the request should be tried again. Otherwise, ``RetryConfiguration/default`` configuration is used.
+
+```swift
+let retryConfiguration = RetryConfiguration(retries: 2, delay: .constant(1)) { error in 
+    // custom logic here
+}
+let userResponse: UserResponse = try await apiManager.request(
+    UserRouter.getUser,
+    retryConfiguration: retryConfiguration
+)
+```
+
+## DownloadAPIManager
+DownloadAPIManager is responsible for the creation and management of a network file download. It conforms to the ``DownloadAPIManaging`` protocol which allows you to define your own custom DownloadAPIManager if needed. Multiple parallel downloads are supported.
+
+The initialisation is equivalent to APIManager, except the session is created for the user based on a given `URLSessionConfiguration`:
+```swift
+init(
+    urlSessionConfiguration: URLSessionConfiguration = .default,
+    requestAdapters: [RequestAdapting] = [],
+    responseProcessors: [ResponseProcessing] = [StatusCodeProcessor.shared],
+    errorProcessors: [ErrorProcessing] = []
+)
+```
+
+Adapters and processors are passed during initialisation and cannot be changed afterwards.
+
+The DownloadAPIManager contains a public property that enables you to keep track of current tasks in progress.
+```swift
+var allTasks: [URLSessionDownloadTask] { get async }
```

Comment:
when seeing this, thinking about better naming, is it equivalent to runningTasks? or does this array contain different  semantic?

Reply:
agreed, I wouldn't use running as the tasks might be paused, but I'd change it to activeTasks